skops/cli/_update.py

```diff
@@ -0,0 +1,131 @@
+from __future__ import annotations
+
+import argparse
+import json
+import logging
+import zipfile
+from pathlib import Path
+
+from skops.cli._utils import get_log_level
+from skops.io import dump, load
+from skops.io._protocol import PROTOCOL
+
+
+def _update_file(
+    input_file: str | Path,
+    output_file: str | Path | None = None,
+    inplace: bool = False,
+    logger: logging.Logger = logging.getLogger(),
+) -> None:
+    """Function that is called by ``skops update`` entrypoint.
+
+    Loads a skops model from the input path, updates it to the current skops format, and
+    saves to an output file. It will overwrite the input file if `inplace` is True.
+
+    Parameters
+    ----------
+    input_file : str, or Path
+        Path of input skops model to load.
+
+    output_file : str, or Path, default=None
+        Path to save the updated skops model to.
+
+    inplace : bool, default=False
+        Whether to update and overwrite the input file in place.
+
+    logger : logging.Logger, default=logging.getLogger()
+        Logger to use for logging.
+    """
+    if inplace:
+        if output_file is None:
+            output_file = input_file
+        else:
+            raise ValueError(
+                "Cannot specify both an output file path and the inplace flag. Please"
+                " choose whether you want to create a new file or overwrite the input"
+                " file."
+            )
+
+    input_model = load(input_file, trusted=True)
+    with zipfile.ZipFile(input_file, "r") as zip_file:
+        input_file_schema = json.loads(zip_file.read("schema.json"))
+
+    if input_file_schema["protocol"] == PROTOCOL:
+        logger.info(
+            "File was not updated because already up to date with the current protocol:"
+            f" {PROTOCOL}"
+        )
+        return None
+
+    if input_file_schema["protocol"] > PROTOCOL:
+        logger.info(
+            "File cannot be updated because its protocol is more recent than the "
+            f"current protocol: {PROTOCOL}"
+        )
+        return None
+
+    if output_file is None:
+        logger.info(
+            f"File can be updated to the current protocol: {PROTOCOL}. Please"
+            " specify an output file path or use the `inplace` flag to create the"
+            " updated Skops file."
+        )
+        return None
+
+    dump(input_model, output_file)
```

Comment:
This is an extreme edge case, but I wanted to bring it up: If the user cancels the process during the `dump`ing operation, could they end up with a broken file? Since we write to the zip archive progressively, I think it could happen. Sometimes, I would start a command and see I did a mistake and would cancel it, so it's not impossible.

Now when the user uses inplace, they could even bust their original file, ending up with no working version. It is an extreme edge case for this to happen, not sure if we want to take measures against it. @adrinjalali WDYT?

To prevent this behavior, we could write the output to a temp file and then move it to `output_file`.

Reply:
Hi both :) is this what you had in mind? https://github.com/skops-dev/skops/pull/343/commits/3a30721bbdbcc72fa17abc36624c580c51a49943